File: com/example/truthmoji/truthmoji.go
// TruthMoji implementation in Go
package main

import (
	"fmt"
	"sync/atomic"
)

// TruthMoji encapsulates the emoji conversion functionality
type TruthMoji struct {
	emojiTrue   string
	emojiFalse  string
	usageCount  atomic.Int64 // Thread-safe counter
}

// NewTruthMoji creates a new TruthMoji instance
func NewTruthMoji() *TruthMoji {
	tm := &TruthMoji{
		emojiTrue:  "😊",
		emojiFalse: "😔",
	}
	return tm
}

// Emojify converts a boolean value to an emoji
func (tm *TruthMoji) Emojify(value bool) string {
	tm.usageCount.Add(1) // Thread-safe increment
	
	if value {
		return tm.emojiTrue
	}
	return tm.emojiFalse
}

// GetUsage returns the current usage count
func (tm *TruthMoji) GetUsage() int64 {
	return tm.usageCount.Load() // Thread-safe read
}

// Inspect returns a map of internal state for inspection
func (tm *TruthMoji) Inspect() map[string]interface{} {
	return map[string]interface{}{
		"emojiTrue":  tm.emojiTrue,
		"emojiFalse": tm.emojiFalse,
		"usageCount": tm.GetUsage(),
	}
}

func main() {
	// Create a new TruthMoji instance
	tm := NewTruthMoji()
	
	// Demo functionality
	fmt.Println("=== TruthMoji Demo (Go) ===")
	fmt.Println("true →", tm.Emojify(true))
	fmt.Println("false →", tm.Emojify(false))
	fmt.Println("1==1 →", tm.Emojify(1 == 1))
	fmt.Println("1==0 →", tm.Emojify(1 == 0))
	fmt.Printf("Usage: %d\n", tm.GetUsage())
	
	// Demonstrate inspection
	fmt.Println("Inspect:", tm.Inspect())
}

File: com/example/truthmoji/truthmoji.c
/**
 * TruthMoji implementation in C
 * Demonstrates static variables and modular C programming
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <stdatomic.h>

/* TruthMoji structure */
typedef struct {
    const char* emoji_true;
    const char* emoji_false;
    atomic_int usage_count;
} TruthMoji;

/* Initialize a TruthMoji instance */
void truthmoji_init(TruthMoji* tm) {
    tm->emoji_true = "😊";
    tm->emoji_false = "😔";
    atomic_init(&tm->usage_count, 0);
}

/* Convert boolean to emoji */
const char* truthmoji_emojify(TruthMoji* tm, bool value) {
    /* Thread-safe increment */
    atomic_fetch_add(&tm->usage_count, 1);
    
    return value ? tm->emoji_true : tm->emoji_false;
}

/* Get current usage count */
int truthmoji_get_usage(TruthMoji* tm) {
    return atomic_load(&tm->usage_count);
}

/* Print internal state */
void truthmoji_inspect(TruthMoji* tm) {
    printf("TruthMoji { emoji_true: %s, emoji_false: %s, usage_count: %d }\n",
           tm->emoji_true, tm->emoji_false, truthmoji_get_usage(tm));
}

/* Module-level singleton (static instance) */
static TruthMoji g_instance;

/* Module initialization function (called before main) */
__attribute__((constructor))
static void init_module(void) {
    truthmoji_init(&g_instance);
}

/* Public API functions using the singleton */
const char* emojify(bool value) {
    return truthmoji_emojify(&g_instance, value);
}

int get_usage(void) {
    return truthmoji_get_usage(&g_instance);
}

void inspect(void) {
    truthmoji_inspect(&g_instance);
}

/* Demo program */
int main(void) {
    printf("=== TruthMoji Demo (C) ===\n");
    printf("true → %s\n", emojify(true));
    printf("false → %s\n", emojify(false));
    printf("1==1 → %s\n", emojify(1 == 1));
    printf("1==0 → %s\n", emojify(1 == 0));
    printf("Usage: %d\n", get_usage());
    printf("Inspect: ");
    inspect();
    
    return 0;
}